Cleanup build visit modules

biagas opened the pull request:
### Description

Resolves #20383, parts of #20386.
Removed `gpl` license and tags from `modules.xml`

Reduced tags in `modules.xml` to only use grouped items and limit groups to 
* `required` (built by default, individual libs in this group can be disabled with `--no-<name>`, e.g. `--no-python`), 
* `optional` (built when requested either individually or via `--optional`), and 
* `extras` (only built when individually requested).

Libraries should only be listed in one of the groups.

Reduced code duplication where possible.


### Type of change

* ~~[ ] Bug fix~~
* ~~[ ] New feature~~
* ~~[ ] Documentation update~~
* [X] Other
build system cleanup

### How Has This Been Tested?

I ran build_visit with various options to ensure correct behavior. I used `--download-only` for quicker testing.

`--no-thirdparty --cmake` : only cmake was downloaded
`--no-python`:  everything in required was downloaded except python
`--optional`: all required and all optional were downloaded
`--optional --no-moab --llvm --mesagl`   All required, all optional except moab, llvm and mesagl were downloaded.
`--hdf5 --silo`: all required, hdf5 and silo were downloaded.


### Checklist:

- [X] I have commented my code where applicable.
- [X] I have updated the release notes.
- [X] I have made corresponding changes to the documentation.
- ~~[ ] I have added debugging support to my changes.~~
- ~~[ ] I have added tests that prove my fix is effective or that my feature works.~~
- ~~[ ] I have confirmed new and existing unit tests pass locally with my changes.~~
- ~~[ ] I have added new baselines for any new tests to the repo.~~
- ~~[ ] I have NOT made any changes to [*protocol* or *public interfaces*][3] in an RC branch.~~

biagas committed:
Remove gpl license tag from modules.xml

biagas committed:
Remove advanced tag from modules.xml.

biagas committed:
Reduce groups to 'required', 'optional', and 'extras'.
Libraries in the 'extras' group will only be built if specified with --name, e.g. --mpich.

biagas committed:
Support only three  groups: required, optional, extras.
Reduce code duplication where possible.

biagas committed:
Merge branch 'develop' into task/biagas/cleanup_build_visit_modules

biagas commented:
In draft mode for comment on the new approach.

markcmiller86 commented on src/tools/dev/scripts/bv_support/bv_main.sh:
```diff
@@ -700,11 +699,33 @@ function strip_quotes
+# Modifications:
+#   Kathleen Biagas, Monday May 12, 2025
+#   Modified to test for actual group names and return early if possible.
+#   Don't allow 'extras' group to be 'enabled, print message and exit script.
```
```suggestion
#   Don't allow 'extras' group to be 'enabled'. Print message and exit if so.
```

markcmiller86 commented on src/tools/dev/scripts/bv_support/helper_funcs.sh:
```diff
@@ -256,42 +256,32 @@ function add_extra_commandline_args
+#  Modifications: 
+#    Kathleen Biagas, Monday May 12, 2025
+#    Combined verify_required_module_exists and verify_optional_module_exists
+#    into one methd: verify_module exists to reduce code duplication
```
```suggestion
#    into one method, verify_module_exists, to reduce code duplication
```

markcmiller86 reviewed (commented):
@biagas great work on threading through all this logic! I found a couple of typos I think and proposed minor changes.

One thing though...I think there verbiage `required`, `optional` and now `extras` is still potentially unclear to someone venturing into this code for the first time. Is there a place you could add a handful of comments that captures the meaning and differences between these three lists? 

I am still confused on how *optional* mili is different from *extra* mesagl for example.

biagas commented:
> I am still confused on how _optional_ mili is different from _extra_ mesagl for example.

The main benefit  of the `optional` group is to get most of visit's third party libraries with one command line option: `--optional`.

The items in `extras` aren't always needed and/or wanted: `mesagl` may or may not be desired, `mesagl` and `osmesa` are mutually exclusive.  `extras` can also be used to hold libraries that prove to have problematic builds.  It may be the case that there are items currently in `optional` that should be in `extras` for that very reason.

biagas commented:
> One thing though...I think there verbiage `required`, `optional` and now `extras` is still potentially unclear to someone venturing into this code for the first time. Is there a place you could add a handful of comments that captures the meaning and differences between these three lists?

I am not sure where in the code to place it. I do plan on updating our docs on build_visit to reflect whatever changes get the final approval.  

I did add a few extra comments to _--help_, where the individual library options are listed under their group names.

biagas committed:
Apply suggestions from code review

Co-authored-by: Mark C. Miller <miller86@llnl.gov>

markcmiller86 commented:
> > I am still confused on how _optional_ mili is different from _extra_ mesagl for example.
> 
> The main benefit of the `optional` group is to get most of visit's third party libraries with one command line option: `--optional`.

So, it sounds like `optional` is a short-cut, catch-all for *everything available*. Thats reasonable. And, I think `required` is pretty clear but still worth a one-liner somewhere such as...it is not possible to build an instance of VisIt without items in the `required` list.

> 
> The items in `extras` aren't always needed and/or wanted: `mesagl` may or may not be desired, `mesagl` and `osmesa` are mutually exclusive. `extras` can also be used to hold libraries that prove to have problematic builds. It may be the case that there are items currently in `optional` that should be in `extras` for that very reason.

Well, not to sound too pendantic here but "not always needed or wanted" or "may or may not be desired" sound a lot to me like `optional` though I do understand you want to focus here on the impact of the *build* and maybe NOT the available features once built?

Maybe the concept you are after here is `alternates`? osmes and mesagl certainly are alternates and the only reason for choosing one over the other is probably due to build challenges. I dunno...are there run-time impacts regarding that choice...maybe there are as in without osmesa, you can do only screen-capture saves.

But, I think I recall seeing `mpich` mentioned for `extras` too. Was that mentioned as an alternate sort of thing...maybe as an alternate to a vendor MPI? Or, was it mentioned as an *optional* (as in...build VisIt for serial or parallel)? I wonder if spack defines a nomenclature for this sort of thing?

`extras` just sounds a bit too much like `optional` to me.

biagas commented:
@markcmiller86 I am fine with changing the 'optional' and 'extras' group names to whatever seems appropriate.  I kept 'optional' as that is currently in use and familiar with a lot of developers via the '--optional' flag.  @JustinPrivitera suggested 'bonus' in his ticket when he suggested a need for a new group.

'mpich' may not be wanted as part of the build, as you might want to use system MPI which would be setup via other PAR flags.
Needing 'mesagl' or 'osmesa' are dependent on the system, it is feasible to need neither.

markcmiller86 commented:
> @markcmiller86 I am fine with changing the 'optional' and 'extras' group names to whatever seems appropriate. 

I honestly feel like I don't understand the differences well enough to suggest better names. I know only enough to say the existing names and functionality as I've understood them are still confusing to me. Maybe its just me though.

JustinPrivitera commented:
I suggest `double-optional`.

I don't know if there is a good name. It probably makes sense to have libraries that don't fit into either of the `required` and `optional` categories, so I think we just need to bite the bullet and pick the name that is least offensive to our collective sensibilities.

brugger1 commented:
Great work Kathleen!

I like adding a new category for the additional libraries that should only be built in special circumstances. I think `extras` is as good a name as any. I would not change `optional`. This would require additional documentation changes as well as changes to scripts that drive `build_visit`.

I agree with Mark that adding a small amount of documentation in the file to document the differences between the three lists would be helpful.

biagas commented:
> I agree with Mark that adding a small amount of documentation in the file to document the differences between the three lists would be helpful.

Would expanding the comments related to each group in `modules.xml` be sufficient? 
I already updated the 'usage' text with more information for REQUIRED and OPTIONAL library groups.

biagas commented:
Did we come to a consensus on a better name for the third group than `extras`?

Also, my current changes do not allow for turning off the `required` group. It used be do-able because the `no-thirdparty` group automatically meant that `--no-thirdparty` option would turn off that group.  I could add parsing of `--no-thirdparty` from command line to do the same thing as the group used to.  Thoughts?

cyrush commented:
I think it would be good to preserve `-no-thirdparty`.

`extras` should work, users won't really be using it so I don't think we need to sweat over the name.

biagas committed:
Qwt no longer enabled by default since it is an optional library.
It will be enabled with --optional or --qwt.

biagas committed:
Test if lib is enabled before calling its 'ensure' method.

biagas committed:
Add --no-thirdparty arg. It disables the required group in a similar
fashion to the old 'no-thirdparty' group.
Also removed a few no-longer-used variables.

biagas committed:
Add more verbose comments for each group describing its purpose and how things are enabled/disable for each.

biagas committed:
Add a bit more information in --help.

biagas committed:
Merge branch 'develop' into task/biagas/cleanup_build_visit_modules

biagas committed:
Update bv_xml_parser.sh

biagas committed:
Fix --no-thirdparty so that it actually works.

biagas committed:
Minor whitespace adjustment.

biagas committed:
Add more explanation of the modules.xml file and its groups.

biagas committed:
Add build_visit option changes to release notes.

biagas committed:
Merge branch 'develop' into task/biagas/cleanup_build_visit_modules

biagas committed:
Remove '--uintah' from examples listing '--optional' since the library is now automatically an optional library.

cyrush commented on src/doc/dev_manual/CreatingABuildVisItModule.rst:
```diff
@@ -69,7 +70,13 @@ See ``bv_qt.sh`` as an example of allowing system or also an alternate (already
+The `extras` group is for any library (and its dependencies) that should be explicitly requested.
+This could be libraries that are mutually exclusive such as `mesagl` and `osmesa`, or libraries that may not be needed on many systems such as `mpich`, `xcb` or `xkbcommon`.
+The `extras` group should also be used for any library that has proven to have a problematic build system.
+
```
Great docs! Given this write up -- maybe the name for the group should be `explicit`?

biagas replied:
I like the name change. Will implement.

cyrush commented on src/tools/dev/scripts/bv_support/helper_funcs.sh:
```diff
@@ -256,42 +256,32 @@ function add_extra_commandline_args
+    declare -F "bv_${alib}_is_installed" &>/dev/null || errorFunc "${alib} is_installed not found"
+    declare -F "bv_${alib}_is_enabled" &>/dev/null || errorFunc "${alib} is_enabled not found"
 }
 
```
nice refactor to reduce redundant logic

cyrush commented on src/tools/dev/scripts/bv_support/helper_funcs.sh:
```diff
@@ -917,47 +914,46 @@ function prepare_build_dir
+# ****************************************************************************
 
-function check_required_3rdparty
+function check_3rdparty
```
nice refactor to reduce redundant logic

biagas replied:
Thanks!

cyrush reviewed (dismissed):
Looks good, thanks for working through this and for adding good docs and comments.

biagas committed:
Change 'extras' group to 'explicit'.

biagas committed:
Merge branch 'develop' into task/biagas/cleanup_build_visit_modules

markcmiller86 reviewed (approved):
Great work on this @biagas :muscle:. Its a lot of complicated shell coding. Only thing that would make any "cooler" is if it was writen in Perl :wink:. The new name for the new category of dependency is also an improvement.

biagas commented:
> The new name for the new category of dependency is also an improvement.

Thanks to @cyrush for the new name!
